programs/survival/waystones.sc

```diff
@@ -0,0 +1,310 @@
+        ,
+            _app_message(p, str('Waystone structure requires %d %s to activate',
+                global_settings:'structure_size',global_settings:'structure_material'));
+            sound( 'block.copper.step', pos );
+        );
+    );
+);
+
+__on_player_breaks_block(p, block)->(
+    if(block~'lodestone',
+        pos = pos(block);
+        if(global_waystones:pos,
+            sound( 'minecraft:item.trident.thunder', pos );
+            _app_message(p, str('Removed %s Waystone', global_waystones:pos:'name'));
+            delete(global_waystones:pos); 
+        );
+    );
+);
+
+__on_player_right_clicks_block(p, item, hand, block, face, hitvec) -> (
+    if(block~'lodestone' && !p~'sneaking',
+        dimension = current_dimension();
+        if(global_settings:'enabled_dimensions'~dimension,
+            pos = pos(block);
+            
+            if(global_waystones:pos,
+                if( !_handle_name_tag(item, pos, p, hand) && hand=='mainhand',
+                    uuid = p~'uuid';
+                    _read_player_waypoints(uuid);
+                    _mark_player_waypoint(pos, p~'pos', uuid); 
+                    sound( 'minecraft:block.amethyst_block.hit', block );
+                    _open_waypoint_screen(p, pos, uuid);
+                    _write_player_waypoints(uuid);
+                );
+            ,
+                _app_message(p, str('Place the lodestone near a %d block %s structure to activate',
+                    global_settings:'structure_size',global_settings:'structure_material'));
+            );
+        ,
+            _app_message(p, 'Waystones dont\'t work in this dimension');
+            sound( 'item.trident.return', block );
+        );
+    );
+);
+
+_has_required_structure(pos) -> (
+    if(global_settings:'structure_material' && global_settings:'structure_size', 
+        count = 0;
+        scan(pos, 2,2,2, if( _~(global_settings:'structure_material'), count+=1) );
+        count >= global_settings:'structure_size';
+    ,true);
+);
+
+_has_required_item(p) -> (
+    if(!global_settings:'offering', return(true));
+    slot = inventory_find(p, global_settings:'offering');
+    if(slot,
+        item = inventory_get(p, slot);
+        inventory_set(p, slot, item:1 - 1);
+        true;
+    ,
+        _app_message(p, str('The Waystone requires an %s offering', global_settings:'offering'));
+        sound( 'item.trident.return', p~'pos' );
+        false;
+    );
+);
+
+_warp_player(p, pos, dimension, point)->(
+    if(_has_required_item(p),
+        in_dimension(dimension, 
+            run(str('tp %s %d %d %d', p~'command_name', ...(pos + [0.5, 0, 0.5]) ));
+        );
+        schedule( 2, _(p, pos, dimension)->(
+            in_dimension(dimension, 
+                sound( 'item.trident.thunder', pos );
+                particle('totem_of_undying', p~'pos'+[0,1,0], 100, 1, 0.1);
+            );
+        ), p, pos, dimension);
+        _app_message(p, global_waystones:point:'name');
```

Comment:
Maybe something like "Teleported to %s" would be clearer, just the name looks a bit weird.